File: 0314-binary-tree-vertical-order-traversal/0314-binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        res = []
        min_x = float('inf')
        max_x = float('-inf')
        
        order_items = collections.defaultdict(list)
        
        queue = collections.deque([(0, root)])
        
        while queue:
            x, node = queue.popleft()
            order_items[x].append(node.val)
            
            min_x = min(min_x, x)
            max_x = max(max_x, x)
            
            if node.left:
                queue.append((x - 1, node.left))
            if node.right:
                queue.append((x + 1, node.right))
        
        
        for col in range(min_x, max_x + 1):
            res.append(order_items[col])
        
        return res
        

File: 0314-binary-tree-vertical-order-traversal/0314-binary-tree-vertical-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalOrder = function(root) {
    if (!root) return [];
    
    let minX = Infinity
    let maxX = -Infinity
    const res = [];
    const colValues = new Map();
    
    const queue = [[0, root]];
    
    
    while (queue.length > 0) {
        const [c, node] = queue.shift();
        
        if (!colValues.has(c)) {
            colValues.set(c, []);
        }
        colValues.get(c).push(node.val);
        
        minX = Math.min(minX, c);
        maxX = Math.max(maxX, c);
        
        if (node.left) {
            queue.push([c - 1, node.left])
        }
        
        if (node.right) {
            queue.push([c + 1, node.right])
        }   
    }
    
    for (let c = minX; c <= maxX; c++) {
        if (colValues.has(c)) {
            res.push(colValues.get(c))
        }
    }
    
    return res;
    
    
};